### ISSUE TITLE
differentiable cubify

### ISSUE THREAD

**Issue by haithemb** [facebookresearch/pytorch3d#70]
## 🚀 Feature
Is it possible to make cubify differentiable ?

## Motivation
I used cubify in a custom loss and I am getting this runtime error:
element 0 of tensors does not require grad and does not have a grad_fn

**Comment by gkioxari** [comment]
Hi @haithemb! Cubify is not differentiable, explicitly defined [here](https://github.com/facebookresearch/pytorch3d/blob/master/pytorch3d/ops/cubify.py#L49). This is because there is a [thresholding](https://github.com/facebookresearch/pytorch3d/blob/master/pytorch3d/ops/cubify.py#L110) of probabilities in the implementation. You could actually make it differentiable by writing approximate gradients for it. It's a great exercise, I recommend it!

**Comment by haithemb** [comment]
Thanks a lot for the prompt response. Could you give some hints or point me out to papers regarding approximate gradients for it? My input volume is already binary.

**Comment by gkioxari** [comment]
The voxel inputs are already probabilities. So not sure how using a softmax will help. You could sample from the distribution and do something like gumbel-softmax. Or you could define approximate gradients. I will close this issue for now as it's an open problem, but if you devise a solution I'd be happy to take a look!

**Comment by cajoek** [comment]
Hi!
@gkioxari has this been implemented or do you know if an implementation you can point me to?
@
